### ISSUE TITLE
TSC/TSR: the grand fantabulous time series classification and time series regression super mega hyper bake-off

### ISSUE THREAD

**Issue by fkiraly** [alan-turing-institute/sktime#12]
Bake-off in python toolbox with full set of methods and suite of post-hoc analyses.

Conditional on:
- data loading implemented https://github.com/kiraly-group/sktime/issues/8
- baseline methods and transformer principles implemented https://github.com/kiraly-group/sktime/issues/7
- UEA favourite methods implemented https://github.com/kiraly-group/sktime/issues/10
- grid-tuning implemented https://github.com/kiraly-group/sktime/issues/4

**Comment by TonyBagnall** [comment]
update on this, others feel free to comment as I'm in java world
1. Elastic Ensemble. Jason to update. 
2. TSF. done results recreated, will write it up
3. RISE. done results recreated, will write it up
4. BOSS. done, results recreated, but code could be optimised for speed. 
5. Shapelets. Jason and David to update. 
6. HIVE-COTE. I'm not sure if the ensemble itself has been written? Will be very quick to do so if not. Will need a HIVE-COTE that can read from file also. 
7. Experiments: results are currently generated with my experiments class, need to set it all up with Viktor's orchestration.  This will need to have the capacity at least to post process results file for the diagnostics, or it will be not much use for big experiments.

**Comment by ViktorKaz** [comment]
@TonyBagnall regarding point 7. above

This is the output of the orchestration:

[results.zip](https://github.com/alan-turing-institute/sktime/files/3444232/results.zip)

It was generated by running the [Jupyter Notebook in the examples:](https://github.com/alan-turing-institute/sktime/blob/master/examples/experiment_orchestration.ipynb).

I believe the format of the results output follows UEA's template output file format structure. These files are also stored on the HDD and can be accessed and post processed as users wish. 

Please also check [sktime.experiments.analysis](https://github.com/alan-turing-institute/sktime/blob/master/sktime/experiments/analysis.py). This module provides functionality for running a number of statistical tests on the results generated by the orchestrator. 

What else is required in terms of functionality?

Can you please also clarify:

1. Is the output of the results in the format that you expect to see?
2. What post processing of the results is required?

Can you please also run `sktime.classifiers.ensemble.TimeSeriesForestClassifier` on these three datasets: 
[datasets.zip](https://github.com/alan-turing-institute/sktime/files/3444274/datasets.zip) and see whether you get the same results as in the `results.zip` file I attached earlier. 

I believe that if we get the same results this should be enough to close this issue, no?

**Comment by TonyBagnall** [comment]
thanks Victor, I'll ask James to take a look because he has done all our experimentation and evaluation code.

**Comment by James-Large** [comment]
I'm coming at this without too much context/familiarity etc on this side of the repo and obviously I'm no great authority, so I think I'll spell out the various requirements of e.g. a bakeoff style set of experiments as I see them, and then we'll meet somewhere in the middle with what everyone wants/is feasible/is worthwhile. 


This turned into a bit of an essay, TL;DR: 

**1.1)** _Is the output of the results in the format that you expect to see?_ On a high level yes, but probability distributions would be wanted at minimum (for classification, sktime I recognise is for other tasks too), and more information in general is always good.
**1.2)** _What post processing of the results is required?_ The individual results analysis methods already there look decent, all the standard stuff is there, and any more bespoke stuff can only really be defined in response to a need discovered with results to hand. (After writing:) Actually, pairwise win/draw/loss tables between classifiers can also be useful. And for e.g. average_and_std_error, being able to specify between mean and median (for e.g. average timings) can be useful too 



On reproducing the output of the orchestrator, if we talked just in terms of accs: 

```
Viktor Saved Results              
ArrowHead        = 0.611428571
GunPoint         = 0.906666667
ItalyPowerDemand = 0.951409135

Master Branch Notebook Output    
ArrowHead        = 0.906040
GunPoint         = 0.951362
ItalyPowerDemand = 0.718391           

Master Branch Saved Results      
ArrowHead        = 0.72
GunPoint         = 0.906666667
ItalyPowerDemand = 0.951409135   
```

Dev branch failed, but I reckon now it's because I pulled dev but didn't reinstall it via pip in my env causing weird conflicts, can try again another time if needed

So broadly speaking on those results:
**2.1)** My saved results vs victors: Gunpoint and IPD are same, ArrowHead is off. Probably difference in seed, but coincidentally same results on the other two dsets? Yeah, TSF's init has this default on master: random_state = None and random.seed(random_state). Dont know ins and outs of pythons/numpys random, but presumably that means unseeded. Prefer reproducibility even if seed not explicitly provided by user in my opinion, but that's a codebase-wide discussion/decision
**2.2)** My saved results vs my notebook output: Looks like the accuracies ended up in the wrong rows on the notebook output (the numbers for each of these datasets look reasonable on the saved results), and are all slightly different regardless. The saved acc is computed in results_writing.py via 'from sklearn.metrics import accuracy_score as acc', while for the notebook output it is computed in losses.py via ScoreAccuracy's calculate_per_dataset(self, y_true, y_pred). 

**On the file format:** 

They have the same structure, but some things missing. It's an absolute pig of a class, but see the [ClassifierResults](https://github.com/uea-machine-learning/uea-tsc/blob/dev/src/main/java/evaluation/storage/ClassifierResults.java) class for info on the file structure we have over there (it has a bunch of over bloat in it too, but the file structure is the main thing). Depending on how closely everything wants to be mirrored between python/java (a decision above my pay grade atm), and how deep into the different aspects of performance measuring you guys want to do (timings etc), either mirror that as much as possible or even just switch over to a JSON format with equivalently named fields to be honest

Specifics: 
**3.1)** Probabilities should be retrievable/saved. Even if the classifier itself cannot produce probabilities, e.g. a 1 nearest neighbour, probs should be synthesised as a one-hot vector, 1 being at the pred class index. I recognise this is specific to the classification scenario. Is notionally the same file format intended to be used for regressing/forecasting results saving? Where e.g the same meta information is saved, just that for each different task (supervised/unsupervised classification/clustering/regression/forecasting) the information stored on each prediction differs 
**3.2)** Line1 meta/descriptive info. Most of this is hardly used for anything in reality aside from confirmation to the human experimenter/reader that this file represents the results they are hopefully expecting. The time unit (e.g. nanoseconds) is useful if expecting to read results that recorded times in different units, otherwise if you can always assume that everything is measured in e.g. nanoseconds then no real need. In the java version, if no timeunit is supplied then all timings (if any) in the file are assumed to be milliseconds, for example, because that's just the default unit we measured in. Not entirely required for a 'bakeoffredux'
**3.3)** Line2 classifier info. In java, this is just left as a free space for the classifier to write any information about its state or the build/fit process, e.g. parameter settings or weightings or so on. Mainly useful if e.g. intending to post-process the settings of a tuned classifier from a saved parameters search (1 file per parameterset) or something like that. Or e.g. retrieving the weights computed of some process across datasets/folds for some piece of analysis in a paper. Not entirely required for a 'bakeoffredux'
**3.4)** Line3 summaries/timings etc. For backwards compatibility and to give a human reader a summary of performance at a glance, the accuracy is stored here. As well as the build(fit) time, the test(predict) time, an optional benchmark operation timing for comparing times across hardware, memory(mostly goes unused), the number of classes (see *1), and more recently (and the jdoc was not updated properly for this yet...) train performance estimate method (e.g. cv_10) and time. How much of this you port across is up to how much is needed for future experiments and ultimately how much you have access too. Pretty much all of this is optional in the java version in terms of ability to read/write files, just obviously if some information is missing it limits what analyses you can do on the results later on. Our experiments class in java fills most of this in by default
**3.5)** Prediction lines. We store trueClassVal, predClassVal, , prob[class0], ... prob[classC], , predTime, , predDescription(optional, mostly unused). I'd argue the true/predclass and probs should definitely be there. Individual prediction times would be difficult since in python you're typically predicting in batch mode and getting all of them at once from interfaced/imported classifiers that you might not want to mess with. Storing the total test time on line 3 should be fine though. The individual prediction times are mainly for particular applications/datasets with e.g. a streaming/real time prediction requirement, and so that we can look at variance in prediction times etc. Pred descriptions are entirely optional and up to the classifier/human reader to write/interpret. Intended usage was e.g. for an ensemble to output the predictions of each base classifier if desired for later analysis. 

*1 On number of classes and/or a class mapping, @jasonlines  and @TonyBagnall  you guys were talking about that before right, and having that as a requirement in general for things to line up between classifiers or something?

In general, a good rule of thumb I think is experiments are expensive time wise, but storage space is (hopefully) cheap. Writing out all the information you think you ever might need for whatever reason sounds good, which is why I also reckon just using json files is a perfectly valid option, you can add fields without ordering being important and naming each field doesn't matter too much in terms of the bytes used in storage. That being said, I've filled a couple TBs with results files by being over-zealous so there is obviously a limit depending on how much you're generating

For e.g. 20 classifiers, 128 datasets, 30 folds/resamples, and trainestimates generated, that’s 20*128*30*2 = 153600 files generated, size depends on number of instances and therefore amount of prediction info and how ‘neat’ the probabilities are in terms of double precision, 0.1 is obviously cheaper to store than 0.123423123… But suppose each file averages 10kb, that’s ~1.5gb, should be easily manageable. 


**On the orchestrator/experimental requirements**

Typical way in which we'd define all the experiments for a bake-off style evaluation would be same as it is already in the orchestrator file i.e.:

```
	for each classifier/strategy
		for each dataset
			for train/test in each resample/fold
				optionally estimate performance on train set (through e.g. (nested) cv)
					and write trainFoldX file
				build/fit on full train set
				test/predict on test set
					and write testFoldX file
```

The 'trainFoldX' files aren't strictly needed for a generic bake-off most likely and add A LOT to the total experimentation time if used e.g. 10 fold cv. They're useful when e.g. generating results for classifiers which will themselves be postprocessed to form ensembles etc. later on. Useful to have the option in general though, and all the tools are around

Here, I'll call the 'orchestration' the for loops and other parameters of the set of experiments (e.g. data read/results write locations), which define a set of experiments to run, and a single 'experiment' as an evaluation of a particular classifier on a particular fold/resample of a dataset. 

When running sets of experiments locally they'd be run exactly like that, optionally with each experiment spawning a thread. For large-scale sets of experiments (e.g. bakeoff style), each individual experiment is run on separate cluster nodes. (Incidentally, we have our UEA cluster and code towards that, I don't know if you guys intend to allow options for e.g. hadoop clusters or whatever). If you want to emulate that for practicality, there's two main ways I'd argue to do that: 

**4.1)** The orchestrator is always initialised with and enumerates the full set of classifiers/datasets/folds, but if a particular experiment index (or list of indices?) is passed as an arg only that particular experiment is conducted/written on this run (makes sense with a higher-level orchestration class already in place, and works seamlessly with local experiments too)
**4.2)** The user provides the particular experimental parameters to evaluate on each execution, which is typically how we've done it for a mix of historical(sunkcost)/uea-cluster reasons, e.g. we use a bash script to generate experiments (calls to an experiments jar with all the appropriate paras) to be completed on the cluster, where the bash script effectively acts as the orchestrator

**On the post processed anaylsis** 

Either way at the end with saved results for all classifiers/datasets/folds, generate average scores and ranks (e.g. accuracy, AUC, log likelihood etc. for classification) for tables and statistical summaries/tests for e.g. critical difference diagrams, which the experiments.analysis file seems to have. A nice thing as well is to be able to easily slice down the results and repeat analysis on the reduced sets, e.g. get all datasets with 2 classes, and compare classifiers on those, etc. 

Individual diagrams/extra tests to be desired etc. only really come about when you actually have the results and you want to show something with them of course. Probably leave deep-diving into that functionality for later and create as needed. Things like ROC diagrams are good for individual datasets if something particularly interesting is found, but in bake-off style stuff over many datasets that's not so important, etc. 

Yeah, I've gone a bit ham on writing this up, but now take what you want from that I guess

**Comment by mloning** [comment]
Given the varying results discussed above, I'd really appreciate a few unit tests on the benchmark and evaluation framework, currently none of it is tested as far as I know.

**Comment by TonyBagnall** [comment]
thanks James, will digest that, but my immediate take home is
1) if a classifier has a get_proba, it should be called and the probabilities should be written. If it doesnt, probability of predicted class set to 1, others to 0
2) line 2 we set with get_params equivalent

**Comment by TonyBagnall** [comment]
here are examples of the java output for test.

![image](https://user-images.githubusercontent.com/9594042/62241436-1007e380-b3d1-11e9-8e91-64c5271b3a34.png)

To summarise James's informative but long post, Line 1 is generated by the equivalent of the Orchestrator. I cant actually remember what PREDICTIONS signifies, rest is obvious. 
Line 2 is generated by our equivalent of get_params. Can be anything, default to No parameter Info
Line 3 the orchestrator again. Accuracy, build time (here in NANOSECONDS as indicated in line 1), test time, some other stuff I personally never use. 

Note the output per case MUST be in the same order as they are in the test file.  
iirc correctly Franz des not like probabilistic output, but it is essential for HIVE-COTE which weights distributions, so if we are to use the orchestrator it needs to be there.

**Comment by TonyBagnall** [comment]
The other issue is the trainFoldX.csv output. 
![image](https://user-images.githubusercontent.com/9594042/62242114-6295cf80-b3d2-11e9-800f-8c4cbb32364e.png)

The file format is the same, but the manner of getting it is a little more complex. By default, the predictions are obtained through a 10xCV on the train data. I'm pretty sure we stratify the folds or randomise then map back though, since some of the train folds are aligned, e.g. 
all class 0
all class 1 
etc. This will screw up a naive CV. This needs clarifying from our code, I'll take a look. But it gets more complex than that. Some classifiers have alternative mechanisms for generating train results, for example, bagging and using OOB error and OOB estimates. We have an interface TrainAccuracyEstimate which, if implemented, means the classifier does its own thing for trainFoldX.csv. Note the responsibility is with the classifier, i.e. it can give biased estimates from the train (if it does a param search, for example). The Orchestrator probably should be able to turn off the internal mechanism. 

My next medium research goal is to quantify the actual effect of doing these shortcut estimates on the meta ensemble. But thats not for here.

**Comment by ViktorKaz** [comment]
Thanks for the comments everyone. Some preliminary thoughts:

1. Ok, regarding the unit tests, I will implement a couple.
2. Regarding reproducing the output - I believe the differences are due to the random seed. Can we try this exercise once again by changing the first row of cell 6 of the [example notebook](https://github.com/alan-turing-institute/sktime/blob/master/examples/experiment_orchestration.ipynb) to:

```python
clf = TimeSeriesForestClassifier(n_estimators=10, random_state=1)
```

I get the following results:

| Estimator | Accuracy |
|--------------|--------------|
| ItalyPowerDemand | 0.939597 |
| ArrowHead | 0.955253 |
| GunPoint | 0.724138 |

3. Regarding the output format. I also think that JSON format makes more sense. Are you ok to switch to it or you insist on the UEA csv format? 

4. I will add the probabilities and time needed to complete each experiment. 

5. Regarding @jasonlines pseudo code for the experiment. The orchestrator does not currently have  this optional requirement:
```
optionally estimate performance on train set (through e.g. (nested) cv)
```
Also the sequence of the loops is:
```python
for d in datasets:
  for s in strategy:
    for c in cv_folds:
      fit()
      predict()
```

I have two questions: 
i) Interesting that you estimate the performance on the train set as well. Why is this needed? This can definitely be added to the output file. The way I understand it we simply need to add this line: `strategy.predict(X_train)`, correct?
ii) does it matter for you that the sequence of the loops is not the same? The orchestrator first loops by dataset and next by strategy, whereas you do the inverse.

6. Regarding **4.1)** and **4.2)** of @James-Large comment. I see the point of doing it this way in order to automate the creation of tasks on a cluster. The logic of the current orchestrator class is a bit different. Users initialise it only with the strategies and datasets they want to run but they can obviously initialise it several times and create several parallel tasks on a cluster. I'll experiment with this idea, probably it will not be very difficult to add this functionality. 

7. On the post processed analysis.
Let's first finish the other things above first and will look into this as well :)

**Comment by TonyBagnall** [comment]
hi Viktor, these are not presenting in the right order, ArrowHead should be 0.72.4138, other two cant tell just from the values.

>I have two questions:
>i) Interesting that you estimate the performance on the train set as well. Why is this needed? 

HIVE-COTE is a meta-ensemble that combines TSF, RISE, ST, EE and BOSS, weighting their probabilistic predictions by an estimate of accuracy formed on the train data, thats why we need it,. 

>The way I understand it we simply need to add this line: strategy.predict(X_train), correct?

no, because that is biased (predicting on the same set you built on). We do a separate CV to the final build and store the train separately (see my post above).

**Comment by TonyBagnall** [comment]
@ViktorKaz we can do it in JSON, and it makes sense, but it is a fairly significant change for our java toolkit, for the analysis tools that read it all in, and of pretty low priority for us.

**Comment by ViktorKaz** [comment]
@TonyBagnall  I found the mistake. The predictions are saved on the HDD and are subsequently loaded to calculate the score. However, they are not necessarily loaded in the same order in which the datasets were supplied to the orchestrator. 

I don't think that this is a problem because the method `prediction_errors` returns two objects:

```python
strategy_dict, losses_df = analyse.prediction_errors(metric=ScoreAccuracy())
```
strategy_dict is used for running the statistical tests and does not have the names of the datasets. However, losses_df has the names of the datasets. 

Below are the results. Here is a [link](https://github.com/alan-turing-institute/sktime/blob/orchestration/examples/experiment_orchestration.ipynb) to the corrected example.



|   |   |  loss | std_error| Accuacy |
|----------|--------------|--------|------------|----------|
| GunPoint | TimeSeriesForestClassifier | 0.060403 | 0.019517 | 0.939597 |
| ItalyPowerDemand | TimeSeriesForestClassifier	| 0.044747 | 0.006448 | 0.955253 |
| ArrowHead |TimeSeriesForestClassifier | 0.275862 | 0.033883 | 0.724138 |

Regarding the saving of the results. I understand that the most important requirement is to have the probabilities for each class as well. @mloning I don't think that `predict_proba` is part of `sktime.highlevel.TSCTask` and `sktime.highlevel.BaseTask` interface. Can we add it?

@TonyBagnall @James-Large Regarding using JSON as output format. The design of the experiments class supports different backends for saving and loading the results so this can be added subsequently if required. Let's stick to the UEA format for now. However, do you want to change `sktime.utils.results_writing.write_results_to_uea_format` to support the stuff you need. At the moment you have `second_line='N/A'`. From the example that Tony sent there is a lot of stuff that is written to the second line of the file which apparently if some sort of strategy specific `get_params` method. How do you want to handle this second line, leave it empty or make it uniform across sktime and your Java code?

**Comment by mloning** [comment]
Yes, `predict_proba` could be added to `TSCStrategy`

**Comment by mloning** [comment]
I'll close this for now.
